### PR TITLE
Add JFR GCHeapSummary event implementation

### PR DESCRIPTION
**Pull request by kazukg**
### TL;DR
This pull request allows native applications to emit jdk.GCHeapSummary which is a jfr-event.

Jdk.GCHeapSummary is a jfr-event that has not yet been implemented.
GCHeapSummary is useful for observing changes of usedHeapSize that occur before and after a garbage collection event.

The sample events
```
jdk.GCHeapSummary {
  startTime = 00:37:25.636
  gcId = 7
  when = "Before GC"
  heapSpace = {
    start = 0x00000000
    committedEnd = 0x00000000
    committedSize = 0 bytes
    reservedEnd = 0x00000000
    reservedSize = 0 bytes
  }
  heapUsed = 218.6 MB
}

jdk.GCHeapSummary {
  startTime = 00:37:25.639
  gcId = 7
  when = "After GC"
  heapSpace = {
    start = 0x00000000
    committedEnd = 0x00000000
    committedSize = 0 bytes
    reservedEnd = 0x00000000
    reservedSize = 0 bytes
  }
  heapUsed = 138.0 MB
}
```
We can visually monitor the heap usage of the application using VisualVM or JDK Mission Control.

### Goals
Implement Start Time, gcId(gcEpoch), gcWhen, usedHeapSize in GCHeapSummary.

### Non-Goals
Do not yet implement start, committedEnd, committedSize, reservedEnd, reservedSize in GCHeapSummary.

Because I've been struggling to find out where I can obtain those specific values.
If someone can offer suggestions on how to accomplish those tasks, I am willing to give it a shot.
### Details
I developed it with reference to the implementation of jdk.GarbageCollection.

1. Add GCHeapSummaryEvent, GCHeapSummaryEventSupport.
2. Modified JfrEvent to add `jdk.GCHeapSummary`.
3. Modified GCImpl to call GCHeapSummaryEvent.

Added JfrGCWhen to manage strings like "Before GC" and "After GC". 
1. Added JfrGCWhenSerializer and JfrGCWhens.
2. Modified JfrFeature and JfrType to add those classes.

Test Code
1. Add TestGCHeapSummaryEvent
2. Modified JfrFileParser and GCWhenConstantPoolParser to test JfrType.GCWhen.

I have verified that I can display graphs using VisualVM or JDK Mission Control.

visualvm

JDKMissionControl

### CODE REVIEW

**Comment by fniephaus** [comment]
Could you take a look at this please, @roberttoyonaga?

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEvent.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import org.graalvm.compiler.api.replacements.Fold;
+import org.graalvm.nativeimage.ImageSingletons;
+import org.graalvm.word.UnsignedWord;
+
+
+class JfrGCHeapSummaryEvent {
+
+    public static void emitJfrGCHeapSummaryEventBeforeGC(UnsignedWord gcEpoch,long start,long heapUsed) {
+        if (hasJfrSupport() ) {
+            jfrSupport().emitGCHeapSummaryEventBeforeGC(gcEpoch, start,heapUsed);
```
Comment (roberttoyonaga) [inline substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEvent.java:36]: Can you run the eclipse formatter please? As well as  `mx checkstyle`. I think that could help fix some of the PR gate issues.

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        
+            JfrNativeEventWriter.putLong(data, heapUsed); //  @Unsigned @DataAmount("BYTES")  @Label("Heap Used") @Description("Bytes allocated by objects in the heap")  long heapUsed;
+        
+            JfrNativeEventWriter.endSmallEvent(data);
+        }
+
+    }
+
+    @Uninterruptible(reason = "Accesses a JFR buffer.")
+    public void emitGCHeapSummaryEventAfterGC(UnsignedWord gcEpoch, long start, long heapUsed) {
```
Comment (roberttoyonaga) [inline substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java:90]: If `emitGCHeapSummaryEventAfterGC(UnsignedWord, long, long)` and `emitGCHeapSummaryEventBeforeGC(UnsignedWord, long, long)` are basically the same, why not just make `emitGCHeapSummaryEvent(UnsignedWord, long , long,  JfrGCWhen)` to reduce duplication?

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+
+@AutomaticallyRegisteredFeature
+class JfrGCHeapSummaryEventFeature implements InternalFeature {
+
+    @Override
+    public void beforeAnalysis(BeforeAnalysisAccess access) {
+        if (HasJfrSupport.get()) {
+
+            //ImageSingletons.add(JfrGCWhens.class, new JfrGCWhens());
```
Comment (roberttoyonaga) [inline substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java:127]: I think you can remove this comment.

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        
+            JfrNativeEventWriter.endSmallEvent(data);
+        }
+        
+    }
+}
+
+
+@AutomaticallyRegisteredFeature
+class JfrGCHeapSummaryEventFeature implements InternalFeature {
```
Comment (roberttoyonaga) [inline substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java:121]: Hmm I think this should actually be similar to `JfrGCEventFeature` and only be included if the default SerialGC is being used. See `isInConfiguration`

**Review by roberttoyonaga — COMMENTED** [review]
Hi @kazukg! Thanks for contributing. I've added a few minor comments.

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEvent.java (hunk 1 above)
Comment (kazukg) [reply substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEvent.java:36]: Hi @roberttoyonaga !
Thank you for confirming. I will give it a try.

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java (hunk 2 above)
Comment (kazukg) [reply substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java:90]: I'll do that.

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java (hunk 3 above)
Comment (kazukg) [reply substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java:127]: I'll do that.

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java (hunk 4 above)
Comment (kazukg) [reply substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java:102]: I added isInConfiguration.

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java (hunk 2 above)
Comment (kazukg) [reply substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java:60]: @roberttoyonaga 
Could you review the changes I made to the source code?

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrGCWhens.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    public static JfrGCWhens singleton() {
+        return ImageSingletons.lookup(JfrGCWhens.class);
+    }
+
+    @Platforms(Platform.HOSTED_ONLY.class)
+    public synchronized JfrGCWhen addGCWhen(String when) {
+        int id = whens.length;
+        JfrGCWhen result = new JfrGCWhen(id, when);
+
+        JfrGCWhen[] newArr = Arrays.copyOf(whens, id + 1);
```
[the remark this answers is not in the document]
Comment (kazukg) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrGCWhens.java]: I updated the source code to hardcode the initialization of JfrGCWhens.whens, and added getBeforeGCWhen and getAfterGCWhen methods to allow JfrGCHeapSummaryEventSupport to obtain their values.

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java (hunk 3 above)
Comment (kazukg) [reply substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java]: I removed it.

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEvent.java (hunk 1 above)
Comment (kazukg) [reply substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEvent.java]: I made changes to the source code and ran `mx checkstyle` on the substratevm folder, verifying that there are no errors.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrType.java [hunk 6]
```diff
@@ -44,6 +44,8 @@ public enum JfrType {
     FrameType("jdk.types.FrameType"),
     GCCause("jdk.types.GCCause"),
     GCName("jdk.types.GCName"),
+    GCWhen("jdk.types.GCWhen"),
+    VirtualSpace("jdk.types.VirtualSpace"),
```
Comment (roberttoyonaga) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrType.java:48]: Is this used anywhere?

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+import com.oracle.svm.core.jfr.JfrEvent;
+import com.oracle.svm.core.jfr.JfrGCWhen;
+import com.oracle.svm.core.jfr.JfrGCWhens;
+import com.oracle.svm.core.jfr.JfrNativeEventWriter;
+import com.oracle.svm.core.jfr.JfrNativeEventWriterData;
+import com.oracle.svm.core.jfr.JfrNativeEventWriterDataAccess;
+
+class JfrGCHeapSummaryEventSupport {
+
+    private final JfrGCWhen before;
```
Comment (roberttoyonaga) [inline substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java:46]: Would it be simpler if you didn't cache these `JfrGCWhen`s and instead just called `JfrGCWhens.singleton().getBeforeGCWhen()` and `JfrGCWhens.singleton().getAfterGCWhen()` from `emitGCHeapSummaryEventBeforeGC` and `emitGCHeapSummaryEventAfterGC` respectively?

**Review by roberttoyonaga — COMMENTED** [review]
Thanks for making those changes. I added a couple more minor comments.

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+                                                      // long startTime;
+
+            JfrNativeEventWriter.putLong(data, gcEpoch.rawValue()); // @Label("GC Identifier") int
+                                                                   // gcId;
+            JfrNativeEventWriter.putLong(data, gcWhen.getId()); // @Label("When") String when;
+
+            // VirtualSpace
+            JfrNativeEventWriter.putLong(data, 0L); // start
+            JfrNativeEventWriter.putLong(data, 0L); // committedEnd : ulong
+            JfrNativeEventWriter.putLong(data, 0L); // committedSize : ulong
```
Comment (roberttoyonaga) [inline substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java:86]: Just some thoughts: You could try using `GCImpl.getPolicy().getCurrentHeapCapacity()` to record the `committedSize`. I'm not certain, but `reservedSize` [might not be applicable here](https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/AbstractCollectionPolicy.java#L375), so maybe it makes the most sense to leave `reservedEnd` and `reservedSize` as 0 for now anyway. You listed these as non-goals so perhaps they can be done at a later time instead.

Comment (kazukg) [reply substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java:86]: Thank you for the advice. I am not currently able to try it out but I will do so in a few days based on your recommendation.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrType.java (hunk 6 above)
Comment (kazukg) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrType.java:48]: Indeed, it does seem unnecessary. Initially, I added it without fully understanding it. I will remove it and verify that it does not affect the functionality.

Comment (kazukg) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrType.java]: I removed Definition of Virtual Space from JrfType.

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java (hunk 7 above)
Comment (kazukg) [reply substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java]: I have completed it.

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java (hunk 8 above)
Comment (kazukg) [reply substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java]: I have added code to output the committed size, and I have verified that the committed size is recorded in jdk.GCHeapSummary.

**Comment by roberttoyonaga** [comment]
Thank you for making those changes. This seems good to me now. @christianhaeubl when you have time, can you have a look at this as well to see if it's ready for integration?

**Review by roberttoyonaga — APPROVED** [review]

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+import com.oracle.svm.core.jfr.JfrEvent;
+import com.oracle.svm.core.jfr.JfrGCWhen;
+import com.oracle.svm.core.jfr.JfrGCWhens;
+import com.oracle.svm.core.jfr.JfrNativeEventWriter;
+import com.oracle.svm.core.jfr.JfrNativeEventWriterData;
+import com.oracle.svm.core.jfr.JfrNativeEventWriterDataAccess;
+
+class JfrGCHeapSummaryEventSupport {
+
+    public void emitGCHeapSummaryEventBeforeGC(UnsignedWord gcEpoch, long start, long committedSize, long heapUsed) {
```
Comment (roberttoyonaga) [inline substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java:46]: nit: As a refactoring, you could get rid of `emitGCHeapSummaryEventBeforeGC` and `emitGCHeapSummaryEventAfterGC` and call `emitGCHeapSummaryEvent(UnsignedWord , long , long , long , JfrGCWhen)` directly from `emitJfrGCHeapSummaryEventBeforeGC` and `emitJfrGCHeapSummaryEventAfterGC` (by supplying `JfrGCWhens.singleton().getBeforeGCWhen()` or `JfrGCWhens.singleton().getAfterGCWhen()`.)

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrGCWhens.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+ * questions.
+ */
+package com.oracle.svm.core.jfr;
+
+import org.graalvm.compiler.api.replacements.Fold;
+import org.graalvm.nativeimage.ImageSingletons;
+
+public class JfrGCWhens {
+
+    private JfrGCWhen[] whens = new JfrGCWhen[]{
```
Comment (roberttoyonaga) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrGCWhens.java:33]: nit: I think this can be final

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEvent.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2022, 2022, Oracle and/or its affiliates. All rights reserved.
```
Comment (christianhaeubl) [inline substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEvent.java:2]: Please adjust the copyright headers of the files that you added so that they contain the correct year (it seems that you copy/pasted them from other files).

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrGCWhen.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+ * questions.
+ */
+package com.oracle.svm.core.jfr;
+
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
+
+import com.oracle.svm.core.Uninterruptible;
+
+public class JfrGCWhen {
```
Comment (christianhaeubl) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrGCWhen.java:33]: Please use an `enum` instead of a `class`. Then, you also don't need the class `JfrGCWhens`.

**Review by christianhaeubl — COMMENTED** [review]
Thanks for the PR, I added a few comments.

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEvent.java (hunk 11 above)
Comment (kazukg) [reply substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEvent.java:2]: @christianhaeubl 

Thank you for pointing that out. 
I was copying and pasting the headers from other files.

I have made some modifications to several files, in addition to the ones I added. Do I need to update the copyright headers of those files as well?

Specifically, the files in question are JfrEvent, JfrFeature, JftType, and JftFileParser.

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java (hunk 9 above)
Comment (kazukg) [reply substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java:46]: I'll do that.

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+import com.oracle.svm.core.feature.InternalFeature;
+import com.oracle.svm.core.jfr.HasJfrSupport;
+import com.oracle.svm.core.jfr.JfrEvent;
+import com.oracle.svm.core.jfr.JfrGCWhen;
+import com.oracle.svm.core.jfr.JfrGCWhens;
+import com.oracle.svm.core.jfr.JfrNativeEventWriter;
+import com.oracle.svm.core.jfr.JfrNativeEventWriterData;
+import com.oracle.svm.core.jfr.JfrNativeEventWriterDataAccess;
+
+class JfrGCHeapSummaryEventSupport {
```
[the remark this answers is not in the document]
Comment (kazukg) [reply substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java:44]: Thank you for providing a specific example.
I will refer to the implementation of ThreadParkEvent.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrGCWhen.java (hunk 12 above)
Comment (kazukg) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrGCWhen.java:33]: I'll do that.

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEvent.java (hunk 11 above)
Comment (christianhaeubl) [reply substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEvent.java:2]: Regarding the existing files that you modified: you can update the copyright headers if you want to but you don't need to.

Comment (kazukg) [reply substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEvent.java:2]: Thank you for reply.
I only modify the copyright headers of the files that I've added.

Comment (kazukg) [reply substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEvent.java]: I have completed it.

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (kazukg) [reply substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrGCHeapSummaryEventSupport.java]: I have modified JfrGCHeapSummaryEvent to be composed solely of static methods.

**Comment by christianhaeubl** [comment]
Thanks, I did further cleanups on top, see https://github.com/oracle/graal/pull/6705. I also changed how the used/committed heap size is queried so that the logic is more consistent with other places where we are already querying that information. Let me know if you see any issues with those changes. This PR should get merged in the next few days.
